Add an audit event reference generator

ptgott opened the pull request:
Closes #5044
Closes #10350

This change adds a generator for audit event reference documentation based on Web UI audit event formatters and, if available, test fixtures.

Previous attempts to generate an audit event reference drew from the Teleport source using `AuditEvent` declarations (#13615) and naming conventions (#38344), but inconsistencies within the source meant that the resulting generator was inadequate. The Web UI source files represent the most complete set of source data we have on audit event schemas.

Implementation details:

- If an event includes a fixture, use that as an example of its schema. For any event formatter with no test fixture, include the event, type, code, and description in the reference guide using data from the formatter, but don't include an example.
- If one event type includes multiple possible codes, document each code in an H3 section.
- Ignore instances of the same code beyond the first occurrence.

ptgott committed:
Add an audit event reference generator

Closes #5044
Closes #10350

This change adds a generator for audit event reference documentation
based on Web UI audit event formatters and, if available, test fixtures.

Previous attempts to generate an audit event reference drew from the
Teleport source using `AuditEvent` declarations (#13615) and naming
conventions (#38344), but inconsistencies within the source meant that
the resulting generator was inadequate. The Web UI source files
represent the most complete set of source data we have on audit event
schemas.

Implementation details:

- If an event includes a fixture, use that as an example of its schema.
  For any event formatter with no test fixture, include the event, type,
  code, and description in the reference guide using data from the
  formatter, but don't include an example.
- If one event type includes multiple possible codes, document each code
  in an H3 section.
- Ignore instances of the same code beyond the first occurrence.

Co-authored-by: Rafał Cieślak <rafal.cieslak@goteleport.com>

ravicious committed:
Get rid of `any` in types

The events map did not define what types it stores, so TypeScript
defaulted it to Map<any, any>. After adjusting this type, it turned out
that there were many functions with worked on Event but were actually
given ReferencePageEventData, so I changed their signatures too.

Types usually start with upper case, hence ReferencePageEventData and
not referencePageEventData.

The `as` in `Object.keys(formatters) as Array<keyof Formatters>` was
not necessary, as TypeScript correctly infers the type here.

ravicious committed:
Add more context to the README

ravicious committed:
Add shorthand command to build and gen reference

ravicious commented on web/packages/teleport/package.json:
```diff
@@ -6,7 +6,10 @@
     "build": "pnpm build-wasm && vite build",
-    "build-wasm": "pnpm --filter=@gravitational/shared build-wasm"
+    "build-wasm": "pnpm --filter=@gravitational/shared build-wasm",
+    "event-reference": "pnpm event-reference:build && pnpm event-reference:gen",
```
If you want to make sure that the reference stays up to date as people add or modify events, you'd need to add a CI check similar to the one I added in #57700 for `pnpm process-icons`.

ravicious reviewed (approved):
Looks good TypeScript-wise. It'd be good if someone who knows a thing or two about the audit log could take a look at the final reference.

I allowed myself to improve the types a little bit – there were still some `any` left. I also added some more context to the README and added a single command which both builds and generates the reference. This should make it easier to add the CI check that I mentioned in a comment.

ptgott committed:
Check audit event reference docs in CI

Add Make targets for running and checking the audit event reference
docs. Add a step to the GitHub Actions job that runs Web UI linters to
check the audit event reference docs.

zmb3 reviewed (approved):
LGTM. Once this ships we can have a pass at updating the `desc` fields to be more descriptive.
